### ISSUE TITLE
OSCORE Context lifetime.

### ISSUE THREAD

**Issue by sbernard31** [eclipse/leshan#1230]
This issue aims to discuss about Oscore Context lifetime.

In a general way but also mainly what should be targeted for a minimal viable feature.

Here is some information about that : https://github.com/eclipse/leshan/pull/1212#issuecomment-1074932429.

At Leshan side, there is 2 ideas : 
1. Remove the context when an OSCORE security info is removed from SecurityStore. That means we should be able to remove context by Recipient ID).
2. Remove context on device Deregistration. (here we probably also need to be able to remove context by Recipient ID)

About 2), This will be a pretty much different behavior than with DTLS because currently DTLS connection lifetime is not related to LWM2M registration lifetime. So don't know if we should do the same for OSCORE (I mean not linked registraiton lifetime to OSCORE context lifetime ?)

Maybe more at californium side but :
Even with this cleaning mechanism, I have some concern about not limiting the size of the OscoreCtxtDB or the lifetime of Context itself. @rikard-sics do you think this is justified concerned ? do you think this should be part of the minimal viable feature ?

**Comment by rikard-sics** [comment]
Let me add my post from the other issue here too so the discussion is collected:

Yeah, currently there is no specific lifetime of an OSCORE context. The only limit applied is that when the sender sequence number reaches its max value, then the context must not be used further.
https://datatracker.ietf.org/doc/html/rfc8613#section-7.2.1

In the OSCORE ACE profile, if the Token associated to the context expires the context should no longer be used:
https://datatracker.ietf.org/doc/html/draft-ietf-ace-oscore-profile-19#section-6

We are also working on a new draft Key Update for OSCORE (KUDOS) that is mainly about an alternative to the OSCORE appendix B.2 procedure, but it also introduces an explicit expiration parameter for OSCORE contexts:
https://datatracker.ietf.org/doc/html/draft-ietf-core-oscore-key-update-01#section-2.2.1

When it is desirable, contexts can be removed from the DB using removeContext(OSCoreCtx ctx)*. This could be done due to memory restrictions, or if a client has deregistered or been removed by the user in the server UI. If the client wishes to register then a new context will be derived for it (plus running of Appendix B.2).

**Comment by rikard-sics** [comment]
> At Leshan side, there is 2 ideas :
> 
>     1. Remove the context when an OSCORE security info is removed from SecurityStore. That means we should be able to remove context by Recipient ID).
> 
>     2. Remove context on device Deregistration. (here we probably also need to be able to remove context by Recipient ID)

One idea could then be to add a method for deleting a context based on its RID (and ID Context). Or rely on the current way which would be to first retrieve the context and the delete it using _removeContext(OSCoreCtx ctx)_.

**Comment by rikard-sics** [comment]
> About 2), This will be a pretty much different behavior than with DTLS because currently DTLS connection lifetime is not related to LWM2M registration lifetime. So don't know if we should do the same for OSCORE (I mean not linked registraiton lifetime to OSCORE context lifetime ?)

I think the OSCORE context can't really be deleted before the registration lifetime is over, since in such case there would be no context to continue the communication with. To trigger Appendix B.2. for deriving a new context, an existing shared context between the peers is needed. So one way is to make it so that when the context lifetime expires the devices run the Appendix B.2. procedure and then transition to using the new context (one downside is that observations are lost).

**Comment by sbernard31** [comment]
> I think the OSCORE context can't really be deleted before the registration lifetime is over, since in such case there would be no context to continue the communication with. 

But remove the context from the map, does not me when can not create a new context on demand from OscoreParameter storing in OscoreStore.

> one downside is that observations are lost.

I didn't know that ? Is it an OSCORE RFC limitation, a LWM2M limitation or just a current implementation limitation ?

**Comment by rikard-sics** [comment]
> Maybe more at californium side but :
> Even with this cleaning mechanism, I have some concern about not limiting the size of the OscoreCtxtDB or the lifetime of Context itself. @rikard-sics do you think this is justified concerns ? do you think this should be part of the minimal viable feature ? or it's OK to live without it for now ?

If there would be a lifetime, what do you think it should consist of? Number of messages sent, or simply time? As I mentioned above, one way is to regularly run the Appendix B.2. procedure and thus derive a new context to continue the communication with.  Although, for a minimal viable feature (considering it experimental) I think it is okay to live with contexts with no life time.

As for the maximum size of the OscoreCtxDB that should be fairly simply to do by adding a new constructor that can that an integer defining the upper limit of concurrent contexts. So that may be more simple to add already.

**Comment by rikard-sics** [comment]
> But remove the context from the map, does not me when can not create a new context on demand from OscoreParameter storing in OscoreStore.

Yes, that would be possible. But would that not use the exact same security material, thus resulting in the same sender and recipient keys (meaning nonce and key reuse)? (Unless Appendix B.2. is used with this new context.)

**Comment by rikard-sics** [comment]
> I didn't know that ? Is it an OSCORE RFC limitation, a LWM2M limitation or just a current implementation limitation ?

It is not very explicitly stated in the OSCORE RFC except for a note in the related Appendix B.1 procedure:
https://datatracker.ietf.org/doc/html/rfc8613#appendix-B.1.3

During the work on the [KUDOS draft](https://datatracker.ietf.org/doc/html/draft-ietf-core-oscore-key-update-01) for the proposed new OSCORE context rederivation method we spent a lot of time thinking about this.

Fundamentally the problem is that if you keep observations after the current Appendix B.2 has executed you may end up in a situation where an OSCORE protected response message cryptographically matches 2 different requests. It would go something like this:

1. The client starts an observation Obs1 by sending a request Req1 with Partial IV = X
2. The two peers run Appendix B.2, and resets their Sender Sequence Number (SSN) to 0.
3. Later on, while Obs1 is still ongoing, the client sends a new request Req2 also with Partial IV = X. This is not necessarily an observation request.
4. A notification sent by the server for Obs1 or a response to Req2 would both cryptographically match against both Req1 and Req2.

**Comment by rikard-sics** [comment]
Alright, so trying to summarize the points in this thread and my thoughts.

**About HashMapCtxDB**

* Firstly, is there still a need to modify _HashMapCtxDB_ in Californium to allow deleting an OSCORE Security Context by its RID? Currently it can be done as is already done in [OscoreContextCleaner](https://github.com/eclipse-leshan/leshan/blob/780eb563559005ed1d300057e56e216aef6b825c/leshan-tl-cf-server-coap-oscore/src/main/java/org/eclipse/leshan/transport/californium/server/OscoreContextCleaner.java#L86-L90)
* Is there still a need to extend _HashMapCtxDB_ to cap the number of OSCORE Security Contexts? I'd say yes. Or at least make the current count visible.

**Context deletion in general**

As for general rules for deletion of OSCORE Security Contexts I see two cases:
* The client explicitly de-registers: It is fine to remove the OSCORE Security Context in this case, as long as the keying material to re-derive that context is kept.
* The client does not de-register but is "killed" somehow: Thus, there is no explicit trigger to remove the OSCORE Security Context. However, it should still be removed after sufficient time has passed. Here I see three potential sub-cases:
  * The client tries to register again after the server has deleted the context. In such case it's fine for the server to re-derive a new OSCORE Security Context for this client, using the keying material it has stored for the client. As the client is trying to register again it will run Appendix B.2 with the server.
  * Is there a scenario where the client would not try to re-register but send some other kind of request to the server after the server has deleted the context? Maybe not? If there was then in such case it would probably be the server that has to initiate Appendix B.2 context re-derivation. So this case can never happen?
  * If the server wishes to send a request to the client, it can only do that as long as it has kept the OSCORE Security Context for that client, since if the context is deleted the client is considered to have been unregistered. So this case can never happen?
 
**Proposals**

One proposal:
* Delete the OSCORE Security Context when the client explicitly de-registers.
* Delete the OSCORE Security Context for "inactive" clients after a set time period. Potential complication: If the deletion time period is shorter than the registration lifetime, the same problem I describe below where the server has to initiate Appendix B.2 happens.

Another proposal, more in line with how DTLS is handled:
* Always delete an OSCORE Security Context after a set time period. The complication here is that if the client continues communicating with the server, the server could indeed re-derive the OSCORE Security Context for that client from the correct keying material. However, since in this scenario the client has not re-started, it will not be initiating Appendix B.2. Instead it would have to be the server that initiates Appendix B.2 upon reception of the first request* from the client after the server has deleted the OSCORE Security Context. Appendix B.2 can be initiated by either peer, either in a CoAP response or a CoAP request. (*Or alternatively if after the server deletes the OSCORE Security Context it immediately re-derives it and sends a request to the client to initiate Appendix B.2)
  
Additional ideas:
* The client could regularly run Appendix B.2 to continuously refresh contexts. Drawback: Observations will be cancelled.  
    
    
What are your thoughts?

**Comment by sbernard31** [comment]
Thx for very clear thoughts.  🙏 

Before to answer, I need some confirmation.

1. **This is 100% sure that when context "rederived", observations are lost.** ? (you already explain at https://github.com/eclipse-leshan/leshan/issues/1230#issuecomment-1092262201 but I want to be sure because problem is clearly not the same depending of the answer. 🤔 
2. **How this is handled at Californium level ? I mean observation relations are deleted when new contexts is established ?

**Comment by rikard-sics** [comment]
> Thx for very clear thoughts. 🙏
> 
> Before to answer, I need some confirmation.
> 
>     1. **This is 100% sure that when context "rederived", observations are lost.** ? (you already explain at [OSCORE Context lifetime. #1230 (comment)](https://github.com/eclipse-leshan/leshan/issues/1230#issuecomment-1092262201) but I want to be sure because problem is clearly not the same depending of the answer. 🤔
> 
>     2. **How this is handled at Californium level ? I mean observation relations are deleted when new contexts is established ?

Just to acknowledge that I saw the message. I will follow-up with more information.

1. The problem is related to the fact that the _external_aad_ of OSCORE does not include the ID context value. Since Appendix B.2 changes the ID context, which is not captured in the _external_aad_, there is a risk of incorrect cryptographic binding for responses (coming after an execution of Appendix B.2) to requests (that were sent before the execution of Appendix B.2).
2. I need to look into the code of Californium to refresh my memory about how things are handled there.

**Comment by sbernard31** [comment]
I really really don't like observe feature, that's always the one which makes everything more complicated 😞 

-----------------------------------------

To be totally honest I don't understand clearly the rederivation/observation issue.
So : 
- IF you think this is important or could be useful for us that I have a clear understanding of that I propose we do that in another dedicated issue. (Personally, I'm curious but I don't want to take up too much of your time and maybe this is very hard to understand, so could be long ... )
- ELSE we can continue the discussion based on that assumption **when context is "rederived" with Appendix B2, observation relation can not survive**
------------------------------------------
> Since Appendix B.2 changes the ID context ...

And [KUDOS draft](https://datatracker.ietf.org/doc/html/draft-ietf-core-oscore-key-update-01) solves that because ID context is not changed when new security context is established. 

So no more need to tie observation lifetime to security context lifetime ?

If this is true, that means there is another possible way to follow :   give-up with Appendix B.2 to invest in KUDOS instead.
But : 
 - maybe KUDOS is not ready
 - maybe this is a lot of works to implement KUDOS in Californium.
 - LWM2M specification doesn't talk about KUDOS but just Appendix B2 :  _"An implementation SHALL follow the procedure described in Appendix B2 of [OSCORE] when the OSCORE security context is used for the first time."_ 

By the way, If you think that KUDOS is future and at long term there is no reason to use Appendix B2, then we should start to report that to OMA, maybe [creating an issue at OMA](https://github.com/OpenMobileAlliance/OMA_LwM2M_for_Developers/issues).

**Comment by rikard-sics** [comment]
> I really really don't like observe feature, that's always the one which makes everything more complicated 😞
> 
> To be totally honest I don't understand clearly the rederivation/observation issue. So :
> 
>     * IF you think this is important or could be useful for us that I have a clear understanding of that I propose we do that in another dedicated issue. (Personally, I'm curious but I don't want to take up too much of your time and maybe this is very hard to understand, so could be long ... )
> 
>     * ELSE we can continue the discussion based on that assumption **when context is "rederived" with Appendix B2, observation relation can not survive**

I think that for now the second option is better. Let us simply determine that observations cannot be kept after Appendix B.2 is used. We could discuss this further, but I'd rather postpone that and it is also complicated as it's in kind of a grey zone where the RFC is not specific, so we may need to bring in other people in the discussion too. My best understanding of the situation is that technically observations could be preserved, but it is not secure to do so.

> 
> > Since Appendix B.2 changes the ID context ...
> 
> And [KUDOS draft](https://datatracker.ietf.org/doc/html/draft-ietf-core-oscore-key-update-01) solves that because ID context is not changed when new security context is established.
> 
> So no more need to tie observation lifetime to security context lifetime ?
> 
> If this is true, that means there is another possible way to follow : give-up with Appendix B.2 to invest in KUDOS instead. But :
> 
>     * maybe KUDOS is not ready
> 
>     * maybe this is a lot of works to implement KUDOS in Californium.
> 
>     * LWM2M specification doesn't talk about KUDOS but just Appendix B2 :  _"An implementation SHALL follow the procedure described in Appendix B2 of [OSCORE] when the OSCORE security context is used for the first time."_
> 
> 
> By the way, If you think that KUDOS is future and at long term there is no reason to use Appendix B2, then we should start to report that to OMA, maybe [creating an issue at OMA](https://github.com/OpenMobileAlliance/OMA_LwM2M_for_Developers/issues).

Indeed KUDOS would solve this problem, and has an explicit mechanism for preserving observations:
https://datatracker.ietf.org/doc/html/draft-ietf-core-oscore-key-update-09#name-preserving-observations-acr

However, as you list, there are a number of complications with using KUDOS. In my mind these are:
- KUDOS is still only a draft and is in fact undergoing major design changes which will be reflected in the next version
- We'd need an implementation for Californium, aligned with the coming draft content
- I suspect we may need to wait until KUDOS is an RFC to get it integrated into Californium
- KUDOS is not defined for use in LwM2M, although that would be great and I will bring up raising it in OMA with co-authors of KUDOS and other relevant people


So to summarize my view is this
- Observations cannot be maintained after Appendix B.2 is used. That is, after running Appendix B.2 all prior observations must be terminated.
- We stick to using Appendix B.2 and not KUDOS for OSCORE in Leshan.

Of course we can re-visit this in the future when things have changed. I suspect that switching Appendix B.2 for KUDOS would not be a too difficult update to do at a later time.

**Comment by sbernard31** [comment]
> - We'd need an implementation for Californium, aligned with the coming draft content
> - I suspect we may need to wait until KUDOS is an RFC to get it integrated into Californium

Maybe this is exactly what you have in mind but just to be sure. I think this could be a good thing that implementation exist when this is still a draft because this way maybe you can get feedback from devs or users. And so this could be easier to fix the RFC when this is still a draft.

> -  Observations cannot be maintained after Appendix B.2 is used. That is, after running Appendix B.2 all prior observations must be terminated.
> -  We stick to using Appendix B.2 and not KUDOS for OSCORE in Leshan.

Ok Let's go with that in mind.

**Comment by sbernard31** [comment]
**About HashMapCtxDB**

>  -  Firstly, is there still a need to modify HashMapCtxDB in Californium to allow deleting an OSCORE Security Context by its RID? Currently it can be done as is already done in OscoreContextCleaner

I don't remember why we thought we need that.
That could be just to enhance API  from user point of view (not so strong argument)
But also maybe that should be because we want it to be done in an atomic way ? (to avoid race condition, I'm not sure there is an issue here)

>  - Is there still a need to extend HashMapCtxDB to cap the number of OSCORE Security Contexts? I'd say yes. Or at least make the current count visible.

If I remember well, we need something like that to write test which ensure there is no memory leak.
(If you have a better idea to achieve that this is OK too)

**Comment by sbernard31** [comment]
This question aims to help me to answer to https://github.com/eclipse-leshan/leshan/issues/1230#issuecomment-2659403010)

**Context lifetime**

So here are facts : 
 - we explicitly say that observations are lost after OSCORE Context rederivation.
 - and knowing that in LWM2M observation are tied to registration. (Meaning that a server is supposed to recreate observation relation after a new registration)


**First Idea : registration lifetime tied to OSCORE context**

A spontaneous idea, could be (at least in a first time,  maybe it could change with KUDOS), to strictly tie registration to OSCORE context. So rederivation means new registration (and vice versa)

This solve most issues  (maybe at cost of more rederivation as strictly needed)
The huge limitation I see : 
  - if server reboot and does not persist OSCoreCtxDB then server can not communicate with client anymore... (it must wait that client re-register again we don't know when 🤷 )....

So we could decide to make possible to persist `OSCoreCtxDB` but this was not the [original idea.](https://github.com/eclipse-leshan/leshan/issues/920) 

**Second Idea : registration lifetime not tied to OSCORE context**

In that case : 
 - server should be able to rederivation context when it lost it. Is it possible right ? 
 - server should have specific behavior for OSCORE and so when observation are lost (after rederivation), it should recreate observe relation. (currently up to Leshan Library  user to do that manually)
 
Probably more work of housekeeping.

**Any opinion ?**

Without digging more (especially the second idea which probably lead to more housekeeping work).
Do you have any opinion ?

**Comment by rikard-sics** [comment]
> **About HashMapCtxDB**
> 
> > * Firstly, is there still a need to modify HashMapCtxDB in Californium to allow deleting an OSCORE Security Context by its RID? Currently it can be done as is already done in OscoreContextCleaner
> 
> I don't remember why we thought we need that. That could be just to enhance API from user point of view (not so strong argument) But also maybe that should be because we want it to be done in an atomic way ? (to avoid race condition, I'm not sure there is an issue here)
>
> > * Is there still a need to extend HashMapCtxDB to cap the number of OSCORE Security Contexts? I'd say yes. Or at least make the current count visible.
> 
> If I remember well, we need something like that to write test which ensure there is no memory leak. (If you have a better idea to achieve that this is OK too)

Yeah I think both these things are worth looking into. 

I think one thing that can be improved is that currently there are 3 internal HashMaps in the _HashMapCtxDB_. Perhaps a design where a context is assigned a unique ID, and the ID is then used in the various HashMaps would be better.

Currently for enforcing a cap, or listing the count of contexts, I'd have to think about which of the HashMaps should be counted. Having an ID would make it easier to understand how many *unique* contexts there are.

The deletion logic could also be improved by having IDs, to be absolutely certain that when a context is deleted, it is in fact removed from all the HashMaps. In a sense keeping the HashMaps synced with each other. 

I will note this down as a TODO and try working on a PR for Californium.

**Comment by rikard-sics** [comment]
> **First Idea : registration lifetime tied to OSCORE context**
> 
> A spontaneous idea, could be (at least in a first time, maybe it could change with KUDOS), to strictly tie registration to OSCORE context. So rederivation means new registration (and vice versa)
> 
> This solve most issues (maybe at cost of more rederivation as strictly needed) The huge limitation I see :
> 
>     * if server reboot and does not persist OSCoreCtxDB then server can not communicate with client anymore... (it must wait that client re-register again we don't know when 🤷 )....
> 
> 
> So we could decide to make possible to persist `OSCoreCtxDB` but this was not the [original idea.](https://github.com/eclipse-leshan/leshan/issues/920)

Yes, I would say that this sounds like the most promising solution, at least for now. Since the observation lifetimes are already tied to the registration this should simplify things.

>     * if server reboot and does not persist OSCoreCtxDB then server can not communicate with client anymore... (it must wait that client re-register again we don't know when 🤷 )....

This issue should be possible to solve by letting the server initiate Appendix B.2 with the client, in case the server reboots. So from a CoAP perspective the server would act as a CoAP client, initiating Appendix B.2 with a request to the client.

**Comment by sbernard31** [comment]
> This issue should be possible to solve by letting the server initiate Appendix B.2 with the client, in case the server reboots. So from a CoAP perspective the server would act as a CoAP client, initiating Appendix B.2 with a request to the client.

Yep but currently Leshan is able to persist observation. So theoretically after a reboot observation are not lost.
But if server need to initiate Appendix B.2  then it should cancel all previous observation. 🤔

**Comment by rikard-sics** [comment]
> > This issue should be possible to solve by letting the server initiate Appendix B.2 with the client, in case the server reboots. So from a CoAP perspective the server would act as a CoAP client, initiating Appendix B.2 with a request to the client.
> 
> Yep but currently Leshan is able to persist observation. So theoretically after a reboot observation are not lost. But if server need to initiate Appendix B.2 then it should cancel all previous observation. 🤔

Right, yeah that is true. So going with the choice of `registration lifetime tied to OSCORE context` we have 2 options as I see it:
1. Persist OSCoreCtxDB 
2. If the server reboots it must use Appendix B.2 with the client, and we have to live with observations being cancelled in such case

Do you have a preference or can think of other options? I suppose the server persisting OSCoreCtxDB is not so unreasonable if we consider that it's most likely a more capable device than the client(s). On the other hand, the server rebooting seems like a rare case (?), where the user may expect some disruptions such as cancelled observations.

**Comment by sbernard31** [comment]
>  I suppose the server persisting OSCoreCtxDB is not so unreasonable if we consider that it's most likely a more capable device than the client(s). 

This breaks a little bit the initial idea : https://github.com/eclipse-leshan/leshan/issues/920
I don't know if this is doable ? Generally the issue about persisting store is how frequent it is to update this store.
(I mean if we need to update it for each messages this could lead to performance issue)
Or maybe try to persist it at shutdown ? (but that doesn't work so well when server is deployed in a cluster)

> On the other hand, the server rebooting seems like a rare case (?), where the user may expect some disruptions such as cancelled observations.

Yep that's true but in my experience production ready server users always expect flawless restart without as much as possible service interruption. 

For a Minimal Viable Feature, this is maybe OK to lost observation and persistence is generally optional feature so default mode should works anyway 🤔. 
My main concerns with that solution is that maybe this will be a bit complicated to handle that correctly. (I mean find the right hook to cancel/remove observation when context is lost)

**Comment by rikard-sics** [comment]
**Regarding the alternative of persisting cancelling/stopping observations on server reboot**:

> For a Minimal Viable Feature, this is maybe OK to lost observation and persistence is generally optional feature so default mode should works anyway 🤔. My main concerns with that solution is that maybe this will be a bit complicated to handle that correctly. (I mean find the right hook to cancel/remove observation when context is lost)
> My main concerns with that solution is that maybe this will be a bit complicated to handle that correctly. (I mean find the right hook to cancel/remove observation when context is lost)

Hmm, what about making it so that for any client using OSCORE the server does not persist observations. What I mean is that currently the server is doing something to store ongoing observations to restore them after reboot. If the server simply did not do that for clients using OSCORE, then observations with such clients would automatically be stopped upon server reboot. Would something like that be feasible?
 
**Regarding the alternative of persisting OSCoreCtxDB**: High-level thought is to take advantage of the procedures described in the OSCORE RFC section B.1.3 to simplify this. See details below.

I don't believe we need to persist everything in the OSCoreCtxDB, in fact I think what would be needed is only the Sender Sequence Number and the ID Context. If the server persists that information it can re-create the same context it had prior to reboot. (As Appendix B.2 changes only the ID Context, that is the key piece of information missing).

As for persisting the Sender Sequence Number, there is a better solution in the OSCORE RFC [here](https://datatracker.ietf.org/doc/html/rfc8613#appendix-B.1.1) that allows for only needing to store it for every X message. So the storing would only have to be done every X messages, or if Appendix B.2 is run (to store the new ID Context). 

Lastly, the server needs to ensure it has a valid replay window, this does not need to be stored on disk but rather the Echo procedure also described in the OSCORE RFC [here](https://datatracker.ietf.org/doc/html/rfc8613#appendix-B.1.2) could be used.

Unfortunately in this case there is also a problem with notification and the server would need to re-register all its past observations at the client as described [here](https://datatracker.ietf.org/doc/html/rfc8613#appendix-B.1.3). Another drawback is that this approach seems more complicated than the one above.

**Summary**

So I see we practically have 3 options:

1. Registration lifetime tied to OSCORE context: Persist OSCoreCtxDB
2. Registration lifetime tied to OSCORE context: Stop observations on server reboot
3. Registration lifetime not tied to OSCORE context

All seem to have some drawbacks and complications. I would probably vote for 2 currently, at least based on our discussion so far.

**Comment by sbernard31** [comment]
**Regarding the alternative of persisting cancelling/stopping observations on server reboot**

>  If the server simply did not do that for clients using OSCORE, then observations with such clients would automatically be stopped upon server reboot. Would something like that be feasible?

I understand the idea and in theory that makes sense.   
But in practice this is not so easy, Observations are store with Registrations in RegistrationStore.
And generally the implementation of stores are either persistent or in memory. 
Having a kind of hybrid (persistent + inMemory) store can bring more complexity.

**Regarding the alternative of persisting OSCoreCtxDB**

Thx for detailed explanation 🙏 

**Summary**

> All seem to have some drawbacks and complications. I would probably vote for 2 (Registration lifetime tied to OSCORE context: Stop observations on server reboot)  currently, at least based on our discussion so far.

Yep Let's start with that.
I don't really encourage to use observation anyway.
This is one more argument to prefer **Send Operation** instead of  **Observation**
And we know that the right solution to solve that in the future is KUDOS.

**Comment by rikard-sics** [comment]
Alright, sounds good! Then we have a plan moving forward.

So I will take the following points as TODOs for Californium:

- Modify HashMapCtxDB to allow deleting an OSCORE Security Context by its RID atomically
- Extend HashMapCtxDB to cap or allow counting the number of OSCORE Security Contexts
- Simplify solution with  3 internal HashMaps in the HashMapCtxDB (identify context by an ID)


Now, when it comes to Leshan, the goal should be to implement `registration lifetime tied to OSCORE context`. To be honest it's not entirely clear to me where to start with this. I suppose it would be about deleting an OSCORE Security Context when the registration for a client is terminated or times out? And additionally only create the OSCORE Security Context in the first place upon client registration (which I believe is already the case?). Basically, what is the delta between the current situation and what we want to achieve? (I feel we're already quite close).

**Comment by rikard-sics** [comment]
Just to structure the steps for implementing `Registration lifetime tied to OSCORE context: Stop observations on server` a bit better than the post above.

We need to ensure the following:

1. Only create the OSCORE Security Context for a client upon client registration
2. Delete the associated OSCORE Security Context when the registration for a client is terminated or times out
3. Make sure that any ongoing observations for clients using OSCORE are terminated after the server reboots (or loses state in general)

Now, my understanding is that points 1 and 2 are fine and is in fact how it works in the current code. In such case, only point 3 would remain to do.

**Comment by sbernard31** [comment]
>  In such case, only point 3 would remain to do.

I read again this conversation and something is not clear to me. 🤔 

What did we decide when the server reboot and we lost OSCORE Context at server side ? 
Is the server : 
 - unable to contact client until the client creates a new context ? 
 - OR it will trigger an appendix B.2 procedure ?

**Comment by rikard-sics** [comment]
> > In such case, only point 3 would remain to do.
> 
> I read again this conversation and something is not clear to me. 🤔
> 
> What did we decide when the server reboot and we lost OSCORE Context at server side ? Is the server :
> 
>     * unable to contact client until the client creates a new context ?
> 
>     * OR it will trigger an appendix B.2 procedure ?

If the server reboots and the OSCORE Security Context with the client is lost, my understanding is that we decided that the server would trigger the Appendix B.2 procedure with the client.

Practically the only way to resolve this kind of deadlock situation is that one of the peers initiates and runs Appendix B.2 with the other. Certainly the client could do this, the problem is that there's no way for the client to understand that the server has rebooted and that running Appendix B.2 is necessary.

Still thinking about our options, if the server reboots, I see the following 2 options (considering the choice of `Registration lifetime tied to OSCORE context: Stop observations on server reboot`):

Option 1: The server can initiate Appendix B.2 with the client.

Option 2: The client will after some time attempt to re-register with the server, now there are 2 possible sub-cases:
  - The server will have no OSCORE Security Context for that client and will not derive one. The server will reply with "Security context not found".
  - The server will have no OSCORE Security Context for that client and will derive one. The server will reply with "Decryption failure". This is because the OSCORE Security Contexts are using different values for the ID Context.

Regardless of which of the 2 sub-cases we fall into, the client could take the response as a hint to attempt to register again, now initiating Appendix B.2. Note that the responses above are unprotected CoAP messages. 

Option 3: The reboot will explicitly trigger sending of a de-register to the client. The client knows it's no longer registered and can register again, now initiating Appendix B.2.

**Comment by sbernard31** [comment]
> Option 2: The client will after some time attempt to re-register with the server, now there are 2 possible sub-cases:

Yep the issue : until the client communicates again with server,  server is no more able to talk to client.
(this is only OK when Queue-mode is used)

So not so good solution.

> Option 3: The reboot will explicitly trigger sending of a de-register to the client

This can be done using 1/x/4 `Disable` resource  (see [E.2 LwM2M Object: LwM2M Server](https://www.openmobilealliance.org/release/LightweightM2M/V1_1_1-20190617-A/HTML-Version/OMA-TS-LightweightM2M_Core-V1_1_1-20190617-A.html#13-2-0-E2-LwM2M-Object-LwM2M-Server))
But if there is a lot of clients to inform that could seriously slow down the server shut-down and this doesn't solve the case where server just die. 

So not so good too.

> Option 1: The server can initiate Appendix B.2 with the client.

OK I want to be sure because for me saying that Registration is tied to OSCORE Context means 1 registration <-> 1 context. (new context -> new registration. new registration -> new context)

But if we do that,  we can not really say that we tie Registration lifetime to OSCORE Context.
Because in case server triggers Appendix B.2, we will have new context but will use same registration.
So in our code we can not assume that : 1 registration <-> 1 context.

> my understanding is that we decided that the server would trigger the Appendix B.2 procedure with the client.

Yep I understood that too but after more thought that sounds not compliant with 1 Registration tied to 1 OSCORE Context.

**My option :** 

I came to the conclusion that maybe we should just tie observations to a context and no more.

I mean we don't care about registration we just do : 
 - when 1 peer (server or client) doesn't have context, it triggers an Appendix B.2.
  - If 1 peer has a context and get a "Security context not found" OR  "Decryption failure", it  triggers an Appendix B.2.

(Like DTLS 👆 )

Now  we need specific code to handle observations : 
 - we should know that an observation is tied to a given context. (is it possible ?)
 - we need an event when a given context is dropped.  (so we can remove observation on that event)
 - When server starts we check all observation-relation and drop the one tied to registration which are using OSCORE. ( which will trigger an event and so up to users to resend observation if needed)

What do you think ?

**Comment by sbernard31** [comment]
(Please note that I will be unavailable for most of the month, so replies may be delayed 😅 )
